crates/ruff_python_formatter/src/expression/mod.rs

```diff
@@ -925,31 +987,57 @@ impl CallChainLayout {
                 // ```
                 ExprRef::Call(ast::ExprCall { func: inner, .. })
                 | ExprRef::Subscript(ast::ExprSubscript { value: inner, .. }) => {
+                    // We preserve these parentheses so don't recurse
+                    // e.g. (a)[0].x().y().z()
+                    //         ^stop here
+                    if is_expression_parenthesized(inner.into(), comment_ranges, source) {
+                        break;
+                    }
+
+                    was_in_call_like = true;
                     expr = ExprRef::from(inner.as_ref());
                 }
                 _ => {
-                    // We to format the following in fluent style:
-                    // ```
-                    // f2 = (a).w().t(1,)
-                    //       ^ expr
-                    // ```
-                    if is_expression_parenthesized(expr, comment_ranges, source) {
-                        attributes_after_parentheses += 1;
-                    }
-
+                    // We count the first call in the chain even
+                    // if it is not part of an attribute, e.g.
+                    //
+                    // f().g()
+                    // ^^^ count this
+                    call_like_count += u32::from(was_in_call_like);
                     break;
                 }
             }
+        }
 
-            // We preserve these parentheses so don't recurse
-            if is_expression_parenthesized(expr, comment_ranges, source) {
-                break;
+        // Observe that `call_like_count >= attributes_after_parentheses`
+        //
+        // Indeed, the latter accumulates only at an attribute expression
+        // with non-parenthesized, call-like value. After that, we
+        // immediately set `was_in_call_like` to `true`. Any execution path
+        // from that code point to the exit or to the next accumulation of
+        // `attribute_after_parentheses` passes through
+        //
+        // ```
+        // call_like_count += u32::from(was_in_call_like);
+        // ```
+        //
+        // and it does so before it could pass through
+        // `was_in_call_like = false`.
+        //
+        // This justifies the name `fluent_layout_more_often_enabled`
+        // used below.
+        if is_fluent_layout_more_often_enabled(context) {
+            if call_like_count + u32::from(first_attr_value_parenthesized) < 2 {
+                CallChainLayout::NonFluent
+            } else {
+                CallChainLayout::Fluent(AttributeState::CallsOrSubscriptsPreceding(call_like_count))
```

Comment:
I'd find it useful to have some examples of when we're hitting those branches. What does `call_like_count + u32::from(first_attr_value_parenthesized) < 2)` mean?